---
title: "Frequently Asked Questions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{FAQ}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## How do I fix the error message:

### "exit status (140)"?

If your Nextflow log contains an error that looks something like this:

```
Caused by:
  Process `WF2:ANNOTATE:annotate (SAMPLE_ID)` terminated with an error exit status (140)
```

This usually indicates that the process ran out of memory (RAM).
Sometimes this will happen for all samples, other times only certain difficult samples will need more RAM.
Try using the R Shiny GUI to increase the requested memory for the process that failed, then rerun the workflow.

### "exit status (127)"?

This indicates that MitoPilot was unable to find a file, directory, or executable. 
Double check that the paths you specified when creating your project are correct. 
Also make sure you're using the correct execution environment. 
This error is commonly caused by using the "local" executor when running MitoPilot on a computing cluster.

## Can I use my own mitogenome assemblies with MitoPilot?

Yes! See the ["Initializing a Project with User Assemblies"](https://smithsonian.github.io/MitoPilot/#initializing-a-project-with-user-assemblies) section of the main page.

## How do I skip processing certain samples?

There are two ways to make MitoPilot skip a set of samples when running the Assembly or Annotate modules.

If the samples successfully finished all steps of the current module, you can lock them.
To do this, select the samples and click the `LOCK` button. 
Locking will also make those samples available for the next module.

Alternatively, instead of locking the samples, you can modify their state.
To do this, select the samples, click on the `STATE` button, then choose `Pre-Assembly (wait)` or `Pre-Annotate (wait)`.
This can be applied to any samples, regardless of whether they completed the current module.

Setting the state to `wait` can help you skip samples that produce unexpected errors and interrupt the workflow.

## Where can I find complete scripts and logs for each process?

MitoPilot copies the most important output files to the `out` directory. 
However, all of the scripts, logs, and outfiles created by Nextflow are saved in the `work` directory. 

The `work` directory is automatically set up by Nextflow and can be difficult to navigate.
To help with this, we provide `NF_work_dir_PROCESS-NAME.txt` files for each sample in the `out` directory. 
These files contain the path to the Nextflow working directory for a specific process and sample.

Note that in the Nextflow working directories, many files are hidden, including scripts and log files.
You can list hidden files on the command line with `ls -a`.

## How can I export information found in the GUI tables?

We have include an R function `MitoPilot::export_db_to_csv`. 
Please keep in mind that the CSV file generated by this function will be a snapshot of the database.
If you go back into MitoPilot and make any changes, you will need to export the database again.

Alternatively, if you are familiar with SQL, you can directly pull the necessary information from the `.sqlite` database located in your project directory.

## How do I restore annotatations after manual editing?

The easiest way to restore the original annotations is by rerunning the sample with the Nextflow `-resume` flag. This will quickly generate clean, unedited annotations from previously completed results in the `work` directory. This approach could be used to fix any mistakes made during manual edits.

Note that rerunning a sample will erase all manual edits and notes.

To rerun specific samples:

- change the state of the target samples to "Ready to Annotate"
- lock all other samples
- make sure your Nextflow command contains the `-resume` flag

## Why does it take so long to edit a start/stop codon position?

Every time you make a manual edit to the start or stop codon position, MitoPilot has to re-do the sequence alignment.
This may take a long time if you have many reference sequences that are a close match to your target gene.

To speed up the process of manual editing, we have included the `Align fewer refs` checkbox. 
Clicking this checkbox limits your alignments to a maximum of five reference sequences (automatically selecting the top BLAST hits).
Note that this will re-run the alignment algorithm, so your alignment results may change.
