<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Initialize new MitoPilot Project — new_project • MitoPilot</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Initialize new MitoPilot Project — new_project"><meta name="description" content="Initialize new MitoPilot Project"><meta property="og:description" content="Initialize new MitoPilot Project"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MitoPilot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/MitoPilot.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/FAQ.html">Frequently Asked Questions</a></li>
    <li><a class="dropdown-item" href="../articles/NMNH-Hydra.html">NMNH Hydra Setup</a></li>
    <li><a class="dropdown-item" href="../articles/NOAA-SEDNA.html">NOAA SEDNA Setup</a></li>
    <li><a class="dropdown-item" href="../articles/custom_dbs.html">Building Custom Databases</a></li>
    <li><a class="dropdown-item" href="../articles/Fish-Mitogenome-Curation.html">Fish Mitogenome Curation</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Smithsonian/MitoPilot/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Initialize new MitoPilot Project</h1>
      <small class="dont-index">Source: <a href="https://github.com/Smithsonian/MitoPilot/blob/main/R/init_project.R" class="external-link"><code>R/init_project.R</code></a></small>
      <div class="d-none name"><code>new_project.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Initialize new MitoPilot Project</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">new_project</span><span class="op">(</span></span>
<span>  path <span class="op">=</span> <span class="st">"."</span>,</span>
<span>  mapping_fn <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mapping_id <span class="op">=</span> <span class="st">"ID"</span>,</span>
<span>  data_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_depth <span class="op">=</span> <span class="fl">2e+06</span>,</span>
<span>  genetic_code <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  executor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"local"</span>, <span class="st">"awsbatch"</span>, <span class="st">"NMNH_Hydra"</span>, <span class="st">"NOAA_SEDNA"</span><span class="op">)</span>,</span>
<span>  container <span class="op">=</span> <span class="st">"macguigand/mitopilot"</span>,</span>
<span>  custom_seeds_db <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_labels_db <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  config <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Rproj <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  force <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>Path to the project directory (default = current working
directory)</p></dd>


<dt id="arg-mapping-fn">mapping_fn<a class="anchor" aria-label="anchor" href="#arg-mapping-fn"></a></dt>
<dd><p>Path to a mapping file. Should be a csv that minimally
includes an `ID` column with a unique identifier for each sample, a `Taxon`
column containing taxonomic information for each sample, and columns
`R1` and `R2` specifying the names of the raw paired read inputs. May include
additional columns with other sample metadata.</p></dd>


<dt id="arg-mapping-id">mapping_id<a class="anchor" aria-label="anchor" href="#arg-mapping-id"></a></dt>
<dd><p>The name of the column in the mapping file that contains
the unique sample identifiers (default = "ID").</p></dd>


<dt id="arg-data-path">data_path<a class="anchor" aria-label="anchor" href="#arg-data-path"></a></dt>
<dd><p>Path to the directory where the raw data is located. Can be
a AWS s3 bucket even if not using AWS for pipeline execution..</p></dd>


<dt id="arg-min-depth">min_depth<a class="anchor" aria-label="anchor" href="#arg-min-depth"></a></dt>
<dd><p>Minimum sequencing depth after pre-processing to proceed
with assembly (default: 2000000)</p></dd>


<dt id="arg-genetic-code">genetic_code<a class="anchor" aria-label="anchor" href="#arg-genetic-code"></a></dt>
<dd><p>Translation table for your organisms. See NCBI website
for more info https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</p></dd>


<dt id="arg-executor">executor<a class="anchor" aria-label="anchor" href="#arg-executor"></a></dt>
<dd><p>The executor to use for running the nextflow pipeline. Must
be one of "local" (default) or "awsbatch", "NMNH_Hydra", or "NOAA_SEDNA".</p></dd>


<dt id="arg-container">container<a class="anchor" aria-label="anchor" href="#arg-container"></a></dt>
<dd><p>The docker container to use for pipeline execution.</p></dd>


<dt id="arg-custom-seeds-db">custom_seeds_db<a class="anchor" aria-label="anchor" href="#arg-custom-seeds-db"></a></dt>
<dd><p>Full path to custom seeds database for GetOrganelle</p></dd>


<dt id="arg-custom-labels-db">custom_labels_db<a class="anchor" aria-label="anchor" href="#arg-custom-labels-db"></a></dt>
<dd><p>Full path to custom labels database for GetOrganelle</p></dd>


<dt id="arg-config">config<a class="anchor" aria-label="anchor" href="#arg-config"></a></dt>
<dd><p>(optional) provide a path to an existing custom nextflow config
file. If not provided a config file template will be created based on the
specified executor.</p></dd>


<dt id="arg-rproj">Rproj<a class="anchor" aria-label="anchor" href="#arg-rproj"></a></dt>
<dd><p>(logical) Initialize and open an RStudio project in the project
directory (default = TRUE). This option has no effect if not running
interactively in RStudio.</p></dd>


<dt id="arg-force">force<a class="anchor" aria-label="anchor" href="#arg-force"></a></dt>
<dd><p>(logical) Force recreating of existing project database and
config files (default = FALSE).</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed as default processing parameters to
`new_db()`</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Devin Leopold.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

