<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Building Custom Databases • MitoPilot</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Building Custom Databases">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MitoPilot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/MitoPilot.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/FAQ.html">Frequently Asked Questions</a></li>
    <li><a class="dropdown-item" href="../articles/NMNH-Hydra.html">NMNH Hydra Setup</a></li>
    <li><a class="dropdown-item" href="../articles/NOAA-SEDNA.html">NOAA SEDNA Setup</a></li>
    <li><a class="dropdown-item" href="../articles/custom_dbs.html">Building Custom Databases</a></li>
    <li><a class="dropdown-item" href="../articles/Fish-Mitogenome-Curation.html">Fish Mitogenome Curation</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Smithsonian/MitoPilot/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Building Custom Databases</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Smithsonian/MitoPilot/blob/main/vignettes/custom_dbs.Rmd" class="external-link"><code>vignettes/custom_dbs.Rmd</code></a></small>
      <div class="d-none name"><code>custom_dbs.Rmd</code></div>
    </div>

    
    
<style>
.alert {
  border-left: 5px solid;
  padding: 10px;
  margin: 10px 0;
  border-radius: 5px;
}
.alert-tip { border-color: #28A745; background-color: #E9F7EF; }
.alert-note { border-color: #007BFF; background-color: #EBF5FF; }
.alert-warning { border-color: #FFC107; background-color: #FFF9E6; }
.alert-danger { border-color: #DC3545; background-color: #F8D7DA; }
strong { font-weight: bold; }
</style>
<div class="section level2">
<h2 id="why-use-a-custom-reference-database">Why use a custom reference database?<a class="anchor" aria-label="anchor" href="#why-use-a-custom-reference-database"></a>
</h2>
<p>Currently, MitoPilot only comes packaged with reference databases for
fishes. If you are working on any other taxonomic group, you will need
to compile databases of mitochondrial sequences for your clade.</p>
</div>
<div class="section level2">
<h2 id="what-parts-of-the-mitopilot-pipeline-use-reference-databases">What parts of the MitoPilot pipeline use reference databases?<a class="anchor" aria-label="anchor" href="#what-parts-of-the-mitopilot-pipeline-use-reference-databases"></a>
</h2>
<ul>
<li>GetOrganelle (Assemble module)</li>
<li>Mitos2 (Assemble module)</li>
<li>BLAST (Assemble and Export modules)</li>
</ul>
</div>
<div class="section level2">
<h2 id="build-custom-databases-for-getorganelle">Build custom databases for GetOrganelle<a class="anchor" aria-label="anchor" href="#build-custom-databases-for-getorganelle"></a>
</h2>
<p>Before proceeding, consider reviewing the GetOrganelle <a href="https://doi.org/10.1186/s13059-020-02154-5" class="external-link">paper</a> and <a href="https://github.com/Kinggerm/GetOrganelle/wiki/FAQ#how-to-assemble-a-target-organelle-genome-using-my-own-reference" class="external-link">documentation</a>
to better understand the required database architecture.</p>
<p>GetOrganelle uses two databases, both in FASTA format:</p>
<ul>
<li>A “seed” database containing complete (or partial) mitochondrial
genomes</li>
<li>A “label” database containing individual mitochondrial gene
sequences</li>
</ul>
<p>There are many different ways to build the GetOrganelle databases. We
have provided the following script to assist with this process.</p>
<div class="section level3">
<h3 id="genbankdownloadutil-sh">
<a href="https://github.com/smithsonian/MitoPilot/blob/main/ref_dbs/getOrganelle/GenBankDownload//GenBankDownloadUtil.sh" class="external-link">GenBankDownloadUtil.sh</a><a class="anchor" aria-label="anchor" href="#genbankdownloadutil-sh"></a>
</h3>
<p>This script will perform a GenBank query for all mitochondrial
records matching your search criteria, download those sequences, and
sort them into GetOrganlle “seed” and “label” databases.</p>
<p>Before proceeding, you will need the following dependencies:</p>
<ul>
<li>
<a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/" class="external-link">Entrez
Direct tools</a> (tested with v22.8)</li>
<li>
<a href="https://www.python.org/downloads/" class="external-link">python</a> (tested with
v3.12.2)</li>
<li>
<a href="https://biopython.org/" class="external-link">biopython</a> (tested with
v1.84)</li>
</ul>
<p>If you are working on the NMNH Hydra cluster, python and biopython
are available as a module. Simply run
<code>module load bio/biopython/1.83</code>.</p>
<p>Download the <a href="https://github.com/smithsonian/MitoPilot/blob/main/ref_dbs/getOrganelle/GenBankDownload//GenBankDownloadUtil.sh" class="external-link">GenBankDownloadUtil.sh</a>
script to the directory where you want to create the custom databases.
You will also need to download <a href="https://github.com/smithsonian/MitoPilot/blob/main/ref_dbs/getOrganelle/GenBankDownload//parseGB.py" class="external-link">parseGB.py</a>
to the same directory.</p>
<p>To run the script, you will need to construct an <a href="https://www.ncbi.nlm.nih.gov/nuccore/advanced" class="external-link">advanced GenBank
query</a>.</p>
<p>For example, if you want to download all starfish mitochondrial
sequences, you could use <code>'"Asteroidea"[Organism]'</code>.</p>
<p>The script can take multiple search terms. For example, use
<code>'"Percidae"[Organism] AND "PRJNA720393"[BioProject]'</code> to
download all percid mitochondrial sequences from a specific
BioProject.</p>
<p>Run the script as follows, providing your custom search terms. Make
sure your full query is in single quotes, each search term is in double
quotes, and the query type in in square brackets.</p>
<p><code>bash GenBankDownloadUtil.sh '"my query"[QueryType]'</code></p>
<p>This may take a while depending on how many GenBank records match
your search terms. If working on a computing cluster, we recommend
running this script as a batch job. Below is an example submission
script for the NMNH Hydra cluster.</p>
<div class="alert alert-note">
<p><strong>Note:</strong> The submission script below assumes that you
have the Entrez Direct tools in your PATH (i.e. these tools can be run
from any directory).</p>
</div>
<pre><code># /bin/sh
# ----------------Parameters---------------------- #
#$ -S /bin/sh
#$ -pe mthread 8
#$ -q sThM.q
#$ -l mres=640G,h_data=80G,h_vmem=80G,himem
#$ -cwd
#$ -j y
#$ -N customGetOrgDBs
#$ -o customGetOrgDBs.log

# script to generate custom seed and label DBs for starfish

# load python and biopython module
module load bio/biopython/1.83 # need python and biopython too

# run the script
bash GenBankDownloadUtil.sh '"Asteroidea"[Organism]'</code></pre>
<p>The script will produce several files:</p>
<ul>
<li>
<code>genbank.gb</code> - GenBank file containing all of the
matching records</li>
<li>
<code>multigene.fasta</code> - FASTA file of sequences that
contained multiple gene records, indicating they are either a partial or
complete mitogenome</li>
<li>
<code>multigene.dedup.fasta</code> - same as
<code>multigene.fasta</code>, but with duplicate sequences removed</li>
<li>
<code>nogene.fasta</code> - FASTA file of mitochondrial sequences
with no annotated genes</li>
<li>
<code>nogene.dedup.fasta</code> - same as <code>nogene.fasta</code>,
but with duplicate sequences removed</li>
<li>
<code>singlegene.fasta</code> - FASTA file of mitochondrial gene
sequences</li>
<li>
<code>singlegene.dedup.fasta</code> - same as
<code>singlelocus.fasta</code>, but with duplicate sequences
removed</li>
</ul>
<p>For GetOrganelle:</p>
<ul>
<li>seed database = <code>multigene.dedup.fasta</code> (plus maybe some
sequences from <code>nogene.dedup.fasta</code>)</li>
<li>label database = <code>singlegene.dedup.fasta</code>
</li>
</ul>
<p>The <code>nogene.fasta</code> file contains un-annotated
mitochondrial sequences or mitochondrial sequences from a non-gene
region, such as the D-loop. Consider manually inspecting these
sequences. You may wish to include some of them in your custom
GetOrganelle seed database.</p>
</div>
<div class="section level3">
<h3 id="inspecting-custom-databases">Inspecting custom databases<a class="anchor" aria-label="anchor" href="#inspecting-custom-databases"></a>
</h3>
<p>Here are a few helpful one-liners to inspect and manipulate FASTA
files.</p>
<p><strong><em>Count the number of sequences in a FASTA
file:</em></strong></p>
<pre><code>grep -c "&gt;" singlelocus.dedup.fasta</code></pre>
<p><strong><em>Generate list of FASTA headers:</em></strong></p>
<pre><code>grep "&gt;" singlelocus.dedup.fasta</code></pre>
<p><strong><em>Generate list of unique gene names:</em></strong></p>
<pre><code>grep "&gt;" singlelocus.dedup.fasta | cut -f1 -d" " | sort | uniq</code></pre>
<p><strong><em>Calculate sequence lengths:</em></strong></p>
<pre><code>cat my_file.fasta | awk '$0 ~ "&gt;" {if (NR &gt; 1) {print c;} c=0;printf substr($0,2,100) "\t"; } $0 !~ "&gt;" {c+=length($0);} END { print c; }'</code></pre>
<p><strong><em>Extract specific sequences from a FASTA file with <a href="https://bioinf.shenwei.me/seqkit/" class="external-link">seqkit</a>:</em></strong></p>
<pre><code># Extract sequences based on names.txt
# names.txt should contain full sequences headers, one per line, but no "&gt;" at start
module load bio/seqkit/2.8.1 # ONLY FOR NMNH HYDRA CLUSTER
seqkit grep -f -n names.txt file.fasta &gt; file_subset.fasta</code></pre>
<p><strong><em>Remove sequences from a FASTA file with <a href="https://bioinf.shenwei.me/seqkit/" class="external-link">seqkit</a>:</em></strong></p>
<pre><code># Remove sequences based on name patterns listed in names.txt
# one pattern per line
module load bio/seqkit/2.8.1 # ONLY FOR NMNH HYDRA CLUSTER
seqkit grep -v -f names.txt file.fasta &gt; file_subset.fasta</code></pre>
<div class="alert alert-note">
<p><strong>Note:</strong> <code>GenBankDownloadUtil.sh</code> will
rename any sequence with no GenBank “product” (protein, tRNA, or rRNA)
as “no_product ACCESSION”. You may wish to remove these sequences, as
they often represent poorly annotated genes.</p>
</div>
</div>
<div class="section level3">
<h3 id="adding-your-own-sequences-to-a-custom-database">Adding your own sequences to a custom database<a class="anchor" aria-label="anchor" href="#adding-your-own-sequences-to-a-custom-database"></a>
</h3>
<p>To use unpublished mitogenomes in your custom GetOrganelle seed
database, you will need to combine multiple FASTA files. GetOrganelle
does not require any specific format for the sequence names in the seed
database.</p>
<p>You can easily combine FASTA files with the Linux <code>cat</code>
command:</p>
<pre><code>cat multigene.dedup.fasta my_mitogenomes.fasta more_mitogenomes.fasta &gt; final_seed_db.fasta</code></pre>
<p>You could also add unpublished individual gene sequences to a custom
GetOrganelle label database in a similar manner.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Devin Leopold.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
