<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Fish Mitogenome Curation • MitoPilot</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Fish Mitogenome Curation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MitoPilot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Smithsonian/MitoPilot/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="Fish-Mitogenome-Curation_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="Fish-Mitogenome-Curation_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><script src="Fish-Mitogenome-Curation_files/reactjson-binding-4.0.0/reactjson.js"></script><script src="Fish-Mitogenome-Curation_files/core-js-2.5.1/shim.min.js"></script><script src="Fish-Mitogenome-Curation_files/react-18.2.0/react.min.js"></script><script src="Fish-Mitogenome-Curation_files/react-18.2.0/react-dom.min.js"></script><script src="Fish-Mitogenome-Curation_files/react-json-view-1.19.1/main.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Fish Mitogenome Curation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Smithsonian/MitoPilot/blob/main/vignettes/Fish-Mitogenome-Curation.Rmd" class="external-link"><code>vignettes/Fish-Mitogenome-Curation.Rmd</code></a></small>
      <div class="d-none name"><code>Fish-Mitogenome-Curation.Rmd</code></div>
    </div>

    
    
<style>
  .reactjson {
    width: auto !important;
    max-width: 100% !important;
    box-sizing: border-box;
}
</style>
<p>Automated curation and validation of mitogenome annotations is a
central goal of MitoPilot. When samples are processed through the
Annotation Module, annotations are first generated using a combination
of <a href="https://gitlab.com/Bernt/MITOS" class="external-link">MITOS2</a> (for rRNA and PCG
annotation) and <a href="https://github.com/UCSC-LoweLab/tRNAscan-SE" class="external-link">tRNAscan-SE</a> (for
tRNA annotation). Following the initial annotation, samples are
proceessed through an automated curations process where modifications
are made in attempt to bring the annotations closer to a state that will
be considered acceptable for publication by NCBI GenBank. This automated
curation process is based on 2 important inputs to the pipeline, a
high-quality reference database of protein coding genes, and a set of
user configurable parameters.</p>
<div class="section level2">
<h2 id="reference-databases">Reference Databases<a class="anchor" aria-label="anchor" href="#reference-databases"></a>
</h2>
<p>MitoPilot expects the curation reference database files to be in the
form of Blastp database files, one for each gene. By default, the
pipeline will use the same database files used by Mitos2 during the
initial annotation process. This can be modified using the
<code>ref_db</code> property of the Curation Parameters (see below). The
default value of ref_db,
(<code>ref_db = list(default = "/ref_dbs/Mitos2/Chordata/featureProt/{gene}.fas")</code>),
specifies the location on the MitoPilot docker image where the database
is located, where <code>{gene}</code> is translated to individual gene
names during processing. This default database location can be
overridden by gene-specific paths added as named entires to the ref_db
parameter,
e.g. <code>ref_db = list(default = "/ref_dbs/Mitos2/Chordata/featureProt/{gene}.fas", atp8 = "/ref_dbs/curation/atp8.fas")</code>.
However, it is important to note that these paths are relative to the
execution environment and must be accessable from that environment.
Building a custom docker image from the base MitoPilot image,
includeding any additional reference databases within its file system,
is one way to achieve this.</p>
</div>
<div class="section level2">
<h2 id="curation-parameters">Curation Parameters<a class="anchor" aria-label="anchor" href="#curation-parameters"></a>
</h2>
<p>Currently, curation parameters are specified at project
initialization and are applied to all samples within the project. By
default, the parameters are set by the function,
<code><a href="../reference/params_fish_mito.html">params_fish_mito()</a></code>. Custimization of the parameters can be
achieved by directly providing a complete named list, or by passing
individual modifications to the <code><a href="../reference/params_fish_mito.html">params_fish_mito()</a></code>
function. For example, the default expected count of trnW genes could be
increased to 2 and the default percent similarity of an “acceptable” PCG
match could be reduced to 85 by initializing a new project with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/new_project.html">new_project</a></span><span class="op">(</span></span>
<span>  <span class="va">...</span>,</span>
<span>  curate_params <span class="op">=</span> <span class="fu"><a href="../reference/params_fish_mito.html">params_fish_mito</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      hit_threshold <span class="op">=</span> <span class="fl">85</span>,</span>
<span>      rules <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        trnW <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>          count <span class="op">=</span> <span class="fl">2</span></span>
<span>        <span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The full set of default curation parameters can be viewed in the
MitoPilot GUI for each sample under “Curation Opts”, and is presented.
At the moments, these values can not be edited for individual samples
from within the GUI, but this feature will eventually be added.
Developer Note: the curation parameters are stored in the project sqlite
database as a base64 encoded text that must be parsed for reading /
writing.</p>
<div class="reactjson html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"data":{"ref_dbs":{"default":"/ref_dbs/Mitos2/Chordata/featureProt/{gene}.fas"},"hit_threshold":90,"max_overlap":0.25,"default_rules":{"rRNA":{"count":1,"max_len":null,"min_len":null,"overlap":{"start":0,"stop":false}},"PCG":{"count":1,"max_len":null,"min_len":null,"overlap":{"start":2,"stop":false},"stop_codons":["TAA","TAG","AGA","AGG","AG","TA","T"],"start_codons":["ATG","GTG","ATA","ATT","TTA","ATC"]},"tRNA":{"count":1,"max_len":null,"min_len":null}},"rules":{"ctrl":{"count":1,"type":"ctrl","min_len":350},"rrnL":{"type":"rRNA","max_len":1850},"rrnS":{"type":"rRNA","max_len":1000},"nad1":{"type":"PCG","start_codons":["ATG","GTG","ATA","ATT","TTA","ATC","TTG"]},"nad2":{"type":"PCG"},"cox1":{"type":"PCG","overlap":{"start":2,"stop":true}},"cox2":{"type":"PCG","start_codons":["ATG","GTG","ATA","ATT","TTA","ATC","TTG"]},"atp8":{"type":"PCG","overlap":{"start":2,"stop":true}},"atp6":{"type":"PCG","overlap":{"start":20,"stop":false},"start_codons":["ATG","GTG","ATA","ATT","TTA","ATC","CTG"]},"cox3":{"type":"PCG"},"nad3":{"type":"PCG"},"nad4l":{"type":"PCG","overlap":{"start":2,"stop":true}},"nad4":{"type":"PCG","overlap":{"start":20,"stop":false}},"nad5":{"type":"PCG","overlap":{"start":2,"stop":true}},"nad6":{"type":"PCG","overlap":{"start":2,"stop":true}},"cob":{"type":"PCG"},"trnA":{"type":"tRNA"},"trnC":{"type":"tRNA"},"trnD":{"type":"tRNA"},"trnE":{"type":"tRNA"},"trnF":{"type":"tRNA"},"trnG":{"type":"tRNA"},"trnH":{"type":"tRNA"},"trnI":{"type":"tRNA"},"trnK":{"type":"tRNA"},"trnL":{"type":"tRNA","count":2},"trnM":{"type":"tRNA"},"trnN":{"type":"tRNA"},"trnP":{"type":"tRNA"},"trnQ":{"type":"tRNA"},"trnR":{"type":"tRNA"},"trnS":{"type":"tRNA","count":2},"trnT":{"type":"tRNA"},"trnV":{"type":"tRNA"},"trnW":{"type":"tRNA"},"trnY":{"type":"tRNA"}}},"name":"Curation Parameters","theme":"monokai","iconStyle":"triangle","indentWidth":4,"collapsed":2,"collapseStringsAfterLength":false,"groupArraysAfterLength":100,"enableClipboard":false,"displayObjectSize":false,"displayDataTypes":false,"onEdit":false,"onAdd":false,"onDelete":false,"onSelect":false,"sortKeys":false},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level2">
<h2 id="validation">Validation<a class="anchor" aria-label="anchor" href="#validation"></a>
</h2>
<p>Following the automated curation process, samples are processed
through a validation script, <code><a href="../reference/validate_fish_mito.html">validate_fish_mito()</a></code> by
default. This script looks for anomolies in the final annotations
relative to the expectations specified in the curation parameters and
add notations to assist with manual curation using the MitoPilot
GUI.</p>
<ul>
<li>
<strong>possible duplicate</strong>: The annotation occurs in the
assembly more often than expected. All identical annotations will
receive this flag and the user must determin which (if any) should be
removed.</li>
<li>
<strong>exceeds max overlap</strong>: The annotation overlaps one or
more neighboring annotations on the same strand by a percentage of its
length greater than the value set by the max_operlap parameter of the
curation params. This represents a global maximum overlap
threshold.</li>
<li>
<strong>exceeds max start overlap</strong>: The start position of a
PCG annotation overlaps one or more neighboring annotations on the same
strand more than expected.</li>
<li>
<strong>exceeds max stop overlap</strong>: The stop position of a
PCG annotation overlaps one or more neighboring annotations on the same
strand more than expected.</li>
<li>
<strong>exceeds max length</strong>: The annotation length exceed
the maximum expectation set in the curation params.</li>
<li>
<strong>below min length</strong>: The annotation length is below
the mimum expectation set in the curation params.</li>
<li>
<strong>low coverage region</strong>: More than 5% of the assembly
region defined by an annotation has less then 10x coverage.</li>
<li>
<strong>high error region</strong>: More than 5% of the assembly
region defined by an annotation has a raw read error rate greater than
5% (ie, more than 5% of raw reads than align to the base show
disagreement on the base call).</li>
<li>
<strong>internal stop codon</strong>: An internal stop codon was
detected in a PCG annotation.</li>
<li>
<strong>non-standard stop codon</strong>: A PCG stop codon other
than those specified in the curation params was detected.</li>
<li>
<strong>non-standard start codon</strong>: A PCG start codon other
than those specified in the curation params was detected.</li>
<li>
<strong>low reference similarity</strong>: The PCG curation
reference database does not include any references with compositional
similarity equal to ro greater than the hit_threshold set in the
curation parameters (default = 90).</li>
<li>
<strong>check reference start alignment</strong>: The PCG start
position does not align exactly with the majority of the top hits from
the reference database.</li>
<li>
<strong>check reference stop alignment</strong>: The PCG stop
position does not align exactly with the majority of the top hits from
the reference database.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Devin Leopold.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
