<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Create a Test Project • MitoPilot</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a Test Project">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MitoPilot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JonahVentures/MitoPilot/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create a Test Project</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/JonahVentures/MitoPilot/blob/main/vignettes/test-project.Rmd" class="external-link"><code>vignettes/test-project.Rmd</code></a></small>
      <div class="d-none name"><code>test-project.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="mitopilot---test-project-tutorial">
<code>{MitoPilot}</code> - Test Project Tutorial<a class="anchor" aria-label="anchor" href="#mitopilot---test-project-tutorial"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>Before running MitoPilot with your own samples, we recommend trying
out the included test project, consisting of Illumina data for 12 fish
species. The following document will walk you through the process of
analyzing this test dataset.</p>
<p>First, make sure you have <a href="https://www.r-project.org/" class="external-link">R
(&gt;=4.0.0)</a> and <a href="https://www.nextflow.io/docs/latest/install.html" class="external-link">Nextflow</a>
installed. This tutorial also assumes that you are using <a href="https://posit.co/download/rstudio-desktop/" class="external-link">RStudio</a> to
interface with R. If you’re working on a computing cluster, we recommend
checking out <a href="https://posit.co/products/open-source/rstudio-server/" class="external-link">RStudio
Server</a>. We have provided detailed instructions for using RStuio
Server on the <a href="SI_NMNH_Hydra_README.html">Smithsonian Hydra</a>
and <a href="NOAA_SEDNA_README.html">NOAA SEDNA</a> computing
clusters.</p>
<p>Next you’ll need to install the <a href="https://github.com/JonahVentures/MitoPilot" class="external-link">MitoPilot</a> R package
from GitHub. Within RStudio, run the following.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"JonahVentures/MitoPilot"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="project-initialization">Project Initialization<a class="anchor" aria-label="anchor" href="#project-initialization"></a>
</h3>
<p>Now we can initialize the test project. For your own data, you would
use the function <code><a href="../reference/new_project.html">MitoPilot::new_project</a></code>. However, for the
test data, we’ll use <code><a href="../reference/new_project.html">MitoPilot::new_project</a></code>.</p>
<pre><code># load the package
library(MitoPilot)

# specify the directory where your test project will be created
wd = "/pool/public/genomics/macguigand/MitoPilot/testing/2024_12_4-6"

# specify an execution environment, "local", "NMNH_Hydra", or "NOAA_SEDNA"
ex = "NMNH_Hydra

# initialize the test project
MitoPilot::new_test_project(
    path = wd, 
    executor = ex, 
    container = "drleopold/mitopilot",
    full_size = FALSE,
    Rproj = FALSE
)</code></pre>
<blockquote>
<p>[!NOTE] UNDER DEVELOPMENT. If you are using an execution environment
that is not currently supported, you could use
<code>config = config.MyEnv</code> to pass a custom Nextflow config to
the <code>new_test_project</code> function.</p>
</blockquote>
<p>If the test project is successfully intialized, you should see the
following.</p>
<pre><code>SRR22396794 - Psychrolutes paradoxus  
SRR22396940 - Psenes pellucidus
SRR22396740 - Hoplostethus occidentalis
SRR21844202 - Fundulus majalis
SRR22396640 - Xyrichtys novacula
SRR22396732 - Gephyroberyx darwinii
SRR22396627 - Gigantura indica
SRR21843972 - Stomias affinis
SRR22396843 - Conger oceanicus
SRR22396668 - Erotelis smaragdus
SRR22396758 - Upeneus parvus
SRR22396865 - Paraconger caudilimbatus
Project initialized successfully.
Please open and review the .config file to ensure all required options are specified.</code></pre>
</div>
<div class="section level3">
<h3 id="exploring-the-mitopilot-gui">Exploring the MitoPilot GUI<a class="anchor" aria-label="anchor" href="#exploring-the-mitopilot-gui"></a>
</h3>
<p>We can now launch the R Shiny Graphical User Interface (GUI) to
examine our test project and start the analysis pipeline.</p>
<pre><code><span><span class="co"># the function to launch the GUI must be called from within your project's directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">wd</span><span class="op">)</span></span>
<span><span class="fu">MitoPilot</span><span class="fu">::</span><span class="fu"><a href="../reference/MitoPilot.html">MitoPilot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<div class="float">
<img src="figures%2Ftest-project%2F1.png" alt="new MitoPilot GUI"><div class="figcaption">new MitoPilot GUI</div>
</div>
<p>Nice! We can see some basic information about our samples. When
intializing your own project, this is pulled from the mapping CSV
file.</p>
<p>You can the samples by any column simply by clicking on the header.
You can also filter samples using the search box in the top right.</p>
<p>To the left of the sample ID column are two icons. The first shows
whether the sample is locked or unlocked. If a sample is unlocked, it
will be included when running the current pipeline stage. A locked
sample cannot be edited for the current stage, but will be made
available for the next stage of the pipeline. You can toggle the lock by
slecting the sample (check mark), then pressing <code>LOCK</code> at the
top of the window. Multiple samples can be locked or unlocked
simultaneously.</p>
<p>The other icon shows the state of the sample. These states vary
depending on the analysis stage and will be automatically updated when
running the pipeline. However, you can also manually change the state of
a sample by selecingt the sample and using the <code>STATE</code>
button.</p>
</div>
<div class="section level3">
<h3 id="modifying-pipeline-parameters">Modifying Pipeline Parameters<a class="anchor" aria-label="anchor" href="#modifying-pipeline-parameters"></a>
</h3>
<p>Within the GUI, we can modify options for each stage of the pipeline.
Select all samples by clicking the empty checkbox at the top left of the
table. Then click on one of the <code>default</code> links in the
<code>Preprocess Opts.</code> column.</p>
<div class="float">
<img src="figures%2Ftest-project%2F3.png" alt="MitoPilot default preprocess"><div class="figcaption">MitoPilot default preprocess</div>
</div>
<p>Here you can see the default options for pre-processing your raw
FASTQ files. We can modify these by clicking the <code>edit</code>
checkbox. Let’s change the CPUs to 4 and the Memory to 40 GB. You could
also change the options passed to <a href="https://github.com/OpenGene/fastp" class="external-link">fastp</a>, but let’s keep those
at the defaults.</p>
<div class="float">
<img src="figures%2Ftest-project%2F4.png" alt="MitoPilot preprocess params"><div class="figcaption">MitoPilot preprocess params</div>
</div>
<p>We can save these new parameter options by clicking on the
<code>Parameter set name</code> box, typing a new name, and clicking
<code>Add YOUR NEW NAME...</code> in the dropdown. Finally, click
<code>Update</code> in the bottom right to save your selection.</p>
<div class="float">
<img src="figures%2Ftest-project%2F5.png" alt="MitoPilot preprocess test params"><div class="figcaption">MitoPilot preprocess test params</div>
</div>
<p>Once you’ve saved a new parameter set, you can easily access again
using the dropdown menu. In the screenshow below, you can see a couple
of other parameter sets that I created.</p>
<div class="float">
<img src="figures%2Ftest-project%2F6.png" alt="MitoPilot preprocess test params"><div class="figcaption">MitoPilot preprocess test params</div>
</div>
<p>After clicking <code>Update</code>, your table should now show
<code>test</code> for all samples in the Preprocess column.</p>
<div class="float">
<img src="figures%2Ftest-project%2F7.png" alt="MitoPilot updated params"><div class="figcaption">MitoPilot updated params</div>
</div>
<p>You can set different parameters for different samples by simply
selecting the set you want to change and repeating the above process.
But for this test dataset, let’s keep everything consistent.</p>
<p>Let’s also change the <code>Assembly Opts.</code>. Repeat the process
above, selecting all samples and creating a new <code>himem</code>
setting.</p>
<div class="float">
<img src="figures%2Ftest-project%2F8.png" alt="MitoPilot asmb params"><div class="figcaption">MitoPilot asmb params</div>
</div>
</div>
<div class="section level3">
<h3 id="runing-mitopilot-assembly">Runing MitoPilot Assembly<a class="anchor" aria-label="anchor" href="#runing-mitopilot-assembly"></a>
</h3>
<p>Let’s get started with the pipeline! Select all samples, then click
the <code>UPDATE</code> button. A new window should appear.</p>
<div class="float">
<img src="figures%2Ftest-project%2F9.png" alt="MitoPilot update asmb"><div class="figcaption">MitoPilot update asmb</div>
</div>
<p>You have two options for running the pipeline. First, you could copy
the Nextflom command and run it within a terminal window on your
computing system. This command could also be imbedded within a
submission script for a computing cluster, if you would prefer. Running
the Nextflow command externally from the GUI may be preferable for
batches with large numbers of samples, since you can perform the
analyses in the background.</p>
<p>Alternatively, you press the <code>Start Nextflow</code> button to
launch the pipeline from within the GUI. Since our test dataset is
small, let’s do that.</p>
<div class="float">
<img src="figures%2Ftest-project%2F10.png" alt="MitoPilot asmb running"><div class="figcaption">MitoPilot asmb running</div>
</div>
<p>Awesome! You should see Nextflow output being printed the to
<code>Progress</code> window. This window will continuely update as the
pipeline progresses. If you’re unsure whether the pipeline is still
running, check to see if the gears in the top right are spinning.</p>
<p>This stage of the pipeline has 3 steps: preprocessing (filtering) the
raw FASTQ files, assembling the mitogenomes using <a href="https://github.com/Kinggerm/GetOrganelle" class="external-link">GetOrganelle</a>, and
calculating coverage maps for the assembly.</p>
<p>The test dataset should take a few minutes to finish. When it’s
done,, your <code>Progress</code> window should look something like
this.</p>
<div class="float">
<img src="figures%2Ftest-project%2F11.png" alt="MitoPilot asmb running"><div class="figcaption">MitoPilot asmb running</div>
</div>
<p>You can now click <code>Close</code> to go back to the samples
table.</p>
</div>
<div class="section level3">
<h3 id="inspecting-assembly-results">Inspecting Assembly Results<a class="anchor" aria-label="anchor" href="#inspecting-assembly-results"></a>
</h3>
<p>The sample table should now be updated with a bunch of new
information.</p>
<div class="float">
<img src="figures%2Ftest-project%2F12.png" alt="MitoPilot asmb results"><div class="figcaption">MitoPilot asmb results</div>
</div>
<p>If you want to locate the results files for a sample, you can scroll
all the way to the right and click <code>output</code>. This will open
the appropriate folder in your RStudio session.</p>
<p>You can also view the results within the MitoPilot GUI. For example,
select SRR21843972 (<em>Stomias affinis</em>), then click
<code>details</code>. This will show you a new window with the
mitogenome sequence.</p>
<div class="float">
<img src="figures%2Ftest-project%2F14.png" alt="SRR21843972 details"><div class="figcaption">SRR21843972 details</div>
</div>
<p>Select that sequence and click the <code>Fasta</code> button in the
bottom right. This will copy the FASTA formatted mitogenome to your
clipboard. This could be useful for a quick BLAST search.</p>
<div class="float">
<img src="figures%2Ftest-project%2F15.png" alt="SRR21843972 fasta"><div class="figcaption">SRR21843972 fasta</div>
</div>
<p>You can also click the <code>view</code> button, which will open a
summary figure in a new tab, including mean read depth, sequence error
rate, and GC content.</p>
<div class="float">
<img src="figures%2Ftest-project%2F16.png" alt="SRR21843972 fig"><div class="figcaption">SRR21843972 fig</div>
</div>
<p>Notice that in this figure, sequence depth drops off dramatically at
both ends of the sequence. SRR21843972 was unable to assemble a circular
mitogenome due to poor read coverage in this region. To try and recover
a circular mitogenome, you could try running the assembly again with
different <code>GetOrganelle</code> settings (see their <a href="https://github.com/Kinggerm/GetOrganelle/wiki/FAQ" class="external-link">Wiki</a>) or
more sequence data.</p>
</div>
<div class="section level3">
<h3 id="problematic-samples">Problematic Samples<a class="anchor" aria-label="anchor" href="#problematic-samples"></a>
</h3>
<p>In this test dataset, there are two problematic samples which
returned failed states. You can see which samples failed by looking for
the exclamation mark state icon. Selecting a sample with that icon and
clicking on the <code>STATE</code> button will cofirm.</p>
<div class="float">
<img src="figures%2Ftest-project%2F13.png" alt="MitoPilot asmb failed state"><div class="figcaption">MitoPilot asmb failed state</div>
</div>
<p>One failed sample is SRR22396758 (<em>Upeneus parvus</em>). We
purposefully truncated the FASTQ file to 200 reads for this sample. With
so few reads, this sample failed to assemble and returned a message
“Insufficient sequencing depth” in the <code>Notes</code> column.</p>
<p>The other failed sample is SRR21844202 (<em>Fundulus majalis</em>).
This sample had plenty of data and was able to assemble a mitogenome.
However, the message in the <code>Notes</code> columns says “Unable to
resolve single assembly from reads”.</p>
<p><a href="https://github.com/Kinggerm/GetOrganelle" class="external-link">GetOrganelle</a>
produced two alternate assembly paths for this sample. This is usually
due to tricky assembly near a repeat region. Please refer to <a href="https://doi.org/10.1186/s13059-020-02154-5" class="external-link">the GetOrganlle
paper</a> for more information about multiple assembly paths.</p>
<p>Let’s take a peek at SRR21844202. Select the sample, scroll all the
way to the right, and click on <code>details</code>.</p>
<div class="float">
<img src="figures%2Ftest-project%2F17.png" alt="SRR21844202 details"><div class="figcaption">SRR21844202 details</div>
</div>
<p>We can see two assemblies listed here. Clicking on <code>view</code>
will show us the coverage, error, and GC content graphs for each
assembly.</p>
<p>Path #1</p>
<div class="float">
<img src="figures%2Ftest-project%2F18.png" alt="path 1"><div class="figcaption">path 1</div>
</div>
<p>Path #2</p>
<div class="float">
<img src="figures%2Ftest-project%2F19.png" alt="path 2"><div class="figcaption">path 2</div>
</div>
<p>The two assembly paths are very similar, differeing only slightly
around 13,000 bp.</p>
<p>We can chose to move forward with just one assembly path and ignore
the other.</p>
<div class="float">
<img src="figures%2Ftest-project%2F22.png" alt="path ignore"><div class="figcaption">path ignore</div>
</div>
<p>Alternatively, we can determine the consensus sequence of both paths.
Selecting both paths and click the <code>Align</code> button in the
bottom right. The sequence alighment will pop up, showing us that the
two paths have 99.9897% sequence similarity.</p>
<p>If we scroll through the alignment, we can see a few bp differences.
For example:</p>
<div class="float">
<img src="figures%2Ftest-project%2F20.png" alt="path align"><div class="figcaption">path align</div>
</div>
<p>Selecting <code>Trim Consensus</code> will remove any conflicting
regions of the aligned assemblies and produce a shorter consensus
sequence of both alignments. Doing so will automatically ignore the
original two assembly paths.</p>
<div class="float">
<img src="figures%2Ftest-project%2F23.png" alt="consensus align"><div class="figcaption">consensus align</div>
</div>
<blockquote>
<p>[!IMPORTANT] Carefully consider your options for problematic samples
with multiple assembly paths. There is no one-size-fits-all
solution.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="runing-mitopilot-annotation">Runing MitoPilot Annotation<a class="anchor" aria-label="anchor" href="#runing-mitopilot-annotation"></a>
</h3>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Devin Leopold.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
